### PR TITLE
logging

### PR DESCRIPTION
**Pull request by Uran198**
So, this is how the logging can look like. It seems that logging may be more dependant on printers than was initially thought, so the printer will probably have to be passed in either way.
Not sure, how much this is better.

### CODE REVIEW

**Comment by Uran198** [comment]
All tests except tests for logging pass.
This is one way of how logging may work. Any thoughts? @coala-analyzer/coala-contributors 
So, if you'll want yourself a logger you'll either create a new one and it will return existing (or create one), or , in future needed loggers will be initialized and configured in the beginning and you'll only call `logging.getLogger` with the appropriate name.

**Comment by Uran198** [comment]
Basically you'll have one logger for every type of `Printer`.

File: coalib/output/printers/LogPrinter.py [hunk 1]
```diff
@@ -1,54 +1,135 @@
...
     """
+    _loggers = {}
+
+    @staticmethod
+    def _get_logger(printer, logger_name, time_fmt=None):
+        """
+        Returns existing or newly created and configured logger.
+        :param logger_name: name of the logger to be returned.
+        :param printer:     printer to initialize logger with if it doesn't
+                            exist, yet.
```
Comment (tushar-rishav) [inline coalib/output/printers/LogPrinter.py]: @Uran198 Forgot to mention about `time_fmt` in docstring  I guess?

Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: @tushar-rishav Yes, for now I'm more interested in general approach. :smile: What's your opinion?

File: coalib/output/printers/LogPrinter.py [hunk 2]
```diff
@@ -1,54 +1,161 @@
+import sys
 import traceback
+import logging
+from colorama import Style
```
Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: It seems to comment on the very wrong line. Don't know if this issue is already reported. @sils1297

Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: It isn't, can you report it to gitlab gitmate coala-incremental?

File: coalib/output/printers/LogPrinter.py [hunk 3]
```diff
@@ -1,54 +1,159 @@
...
-from coalib.output.printers.LOG_LEVEL import LOG_LEVEL, LOG_LEVEL_COLORS
+from coalib.output.printers.LOG_LEVEL import LOG_LEVEL
 from coalib.processes.communication.LogMessage import LogMessage
 
 
+class PrinterHandler(logging.StreamHandler):
+    """
+    Handler for logging which uses printer to print messages and falls back to
+    writing to stdout if no printer is provided.
+    """
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: Can you delete irrelevant comments? Phone ui doesn't let me do it

File: coalib/output/printers/LogPrinter.py [hunk 4]
```diff
@@ -1,54 +1,194 @@
...
+    symbols where appropriate.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        :param use_color:  If True highlights records levelname with respective
+                           color.
+        """
+        self.use_color = kwargs.pop('use_color', False)
+        super(ColoredFormatter, self).__init__(*args, **kwargs)
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: Since we use python 3 IIUC super() suffices?

File: coalib/output/printers/LogPrinter.py [hunk 5]
```diff
@@ -1,54 +1,194 @@
...
+        if not LogPrinter._loggers.get(logger_name):
+            if time_fmt:
+                format_str = '%(levelname)s[%(asctime)s] %(message)s'
+            else:
+                format_str = '%(levelname)s %(message)s'
+
+            use_color = isinstance(printer, ColorPrinter)
+            formatter = ColoredFormatter(format_str,
+                                         time_fmt,
+                                         use_color=use_color)
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: With the formatter we don't need the color printer anymore do we? Also we would want to avoid printing copies when piping strong I guess...

File: coalib/output/printers/LogPrinter.py [hunk 6]
```diff
@@ -1,54 +1,194 @@
...
-                 printer,
+                 printer=None,
                  log_level=LOG_LEVEL.INFO,
-                 timestamp_format="%X"):
+                 timestamp_format="%X",
+                 of_type=True):
         """
-        Creates a new log printer from an existing Printer.
+        Creates a new log printer from an existing Printer or returns existing
+        one if it already exist.
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: So on long term this class will be removed, right? Add deprecation note?

File: coalib/output/printers/LogPrinter.py [hunk 7]
```diff
@@ -1,54 +1,194 @@
...
-                                 LogPrinter, set it to self.
+                                 Printer, set it to self.
         :param log_level:        The minimum log level, everything below will
                                  not be logged.
         :param timestamp_format: The format string for the
                                  datetime.today().strftime(format) method.
+        :param of_type:          If False returns new logger object every time.
+                                 Else may return logger which uses the same
+                                 printer type or will use printer to initialize
+                                 new one.
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: Feels weird, kind of singleton but not really, smells a bit to me. Is it needed or can we just drop the singleton  aspect?

Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: Not sure on this one

Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: The logging works as singletons accessed by name, but coala occasionally uses LogPrinters with StringPrinters, IIRC only in tests, so in time this will change. Basically because of this we need this part, kind of delaying correcting it.

File: coalib/output/printers/LogPrinter.py (hunk 5 above)
Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: Yes, I think we'll be able to fully remove dependence on pyprint in the future. I think colored output was the only reason, why it stayed.

I didn't get the second part...

File: coalib/output/printers/LogPrinter.py (hunk 2 above)
Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: Sure, I'll leave this comments for reference.

File: coalib/output/printers/LogPrinter.py (hunk 6 above)
Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: It may be a little early for the note, cause there is no alternative configured printing solution, yet. Not sure, if it's better to add in this PR configuring logging from files for different kinds of printers.

File: coalib/bears/Bear.py [hunk 8]
```diff
@@ -267,6 +267,9 @@ def __json__(cls):
         Override JSON export of ``Bear`` object.
         """
         _dict = get_public_members(cls)
+        # json cannot serialize properties, so drop them
+        _dict = {key: value for key, value in _dict.items()
```
Comment (sils1297) [inline coalib/bears/Bear.py]: just use `get_public_members(cls).items()`, otherwise you kind of change the meaning of a variable which is worse then just making a new variable

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/bad34a8614f7848783e53d5c2505238b15ac3b95 https://github.com/coala-analyzer/coala/pull/2409/commits/be9acd0e2fac3a5f87b84e417377b08e59b36e2d

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2409/commits/ca195fc5df9a8c195fca2a66c58ae789114c87a8

File: coalib/bears/Bear.py [hunk 9]
```diff
@@ -266,7 +266,9 @@ def __json__(cls):
         """
         Override JSON export of ``Bear`` object.
         """
-        _dict = get_public_members(cls)
+        # json cannot serialize properties, so drop them
```
Comment (sils1297) [inline coalib/bears/Bear.py]: shouldn't the JSONEncoder use the value of the property? It makes a getattr on it which retrieves the value, right?

File: coalib/coala_main.py [hunk 10]
```diff
@@ -52,6 +54,10 @@ def run_coala(log_printer=None,
     :return:                        A dictionary containing a list of results
                                     for all analyzed sections as key.
     """
+    # Configure logging
+    with open('logging_config.json') as fd:
+        logging.config.dictConfig(json.load(fd))
```
Comment (sils1297) [inline coalib/coala_main.py]: can you add a commit adding user documentation for this?

File: coalib/output/printers/LOG_LEVEL.py [hunk 11]
```diff
@@ -1,7 +1,12 @@
...
 
-LOG_LEVEL = enum("DEBUG", "INFO", "WARNING", "ERROR")
-LOG_LEVEL_COLORS = {LOG_LEVEL.ERROR: "red",
-                    LOG_LEVEL.WARNING: "yellow",
-                    LOG_LEVEL.INFO: "blue",
-                    LOG_LEVEL.DEBUG: "green"}
+LOG_LEVEL = enum(DEBUG=logging.DEBUG,
+                 INFO=logging.INFO,
+                 WARNING=logging.WARNING,
+                 ERROR=logging.ERROR)
```
Comment (sils1297) [inline coalib/output/printers/LOG_LEVEL.py]: changing the log level could be easily separately done in an own commit right? (This one's too big IMO.)

File: coalib/output/printers/LOG_LEVEL.py [hunk 12]
```diff
@@ -1,7 +1,12 @@
...
+LOG_LEVEL_COLORS = {LOG_LEVEL.ERROR: Fore.RED,
+                    LOG_LEVEL.WARNING: Fore.YELLOW,
+                    LOG_LEVEL.INFO: Fore.BLUE,
+                    LOG_LEVEL.DEBUG: Fore.GREEN}
```
Comment (sils1297) [inline coalib/output/printers/LOG_LEVEL.py]: this change of using the actual values here could be done as well in an own commit I think

File: coalib/output/printers/LogPrinter.py [hunk 13]
```diff
@@ -1,21 +1,37 @@
...
+
+    def emit(self, record):
+        msg = self.format(record)
+        sys.stdout.write(msg + "\n")
+
+
 class LogPrinter:
     """
+    This class is deprecated and will be soon removed. To get logger use
+    loggin.getLogger('coala.colored') for colored logger and
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: loggin? typo?

File: coalib/bears/Bear.py (hunk 9 above)
Comment (Uran198) [inline coalib/bears/Bear.py]: No, JSONEncoder, gets output of `__json__` which operates on class, so no values of properties here. Then it only filters out some values from a dict and that's it. The only thing that it gets is class attributes, nothing related to instances.

File: coalib/output/printers/LogPrinter.py (hunk 13 above)
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: the user of the logger shouldn't care about colored or not IMO, the coala logger should just not use color if the output is piped e.g. into a file or so

File: logging_config.json [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+        }
+    },
+    "formatters": {
+        "raw": {
+            "format": "[%(levelname)s][%(asctime)s] %(message)s",
+            "datefmt": "%X"
+        }
+    }
+}
+
```
Comment (sils1297) [inline logging_config.json]: a newline too much

File: coalib/coala_main.py (hunk 10 above)
Comment (sils1297) [inline coalib/coala_main.py]: btw. you want to load in a global config or should the user write it? The one below looks global but here youre loading it from current directory

File: coalib/output/printers/LOG_LEVEL.py (hunk 12 above)
Comment (Uran198) [inline coalib/output/printers/LOG_LEVEL.py]: It may be not a good idea to separate this one in separate, cause we will have one commit after which there will be not a working version. Putting this change before or after the commit it is introduced in will break things

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2409/commits/4f8422f569041c4aa49873a5869634260c9e2a50

File: coalib/coala_main.py (hunk 10 above)
Comment (Uran198) [inline coalib/coala_main.py]: It should be global, users shouldn't care about it existence even.

File: coalib/bears/Bear.py (hunk 9 above)
Comment (sils1297) [inline coalib/bears/Bear.py]: why not make `__json__` a method and instead of `key: value` use `key: getattr(self, key)`?

File: coalib/coala_main.py (hunk 10 above)
Comment (Uran198) [inline coalib/coala_main.py]: I'll change that.

As for the docs, what exactly should go there. It's just a file with the configurations, nothing fancy. Or you're saying documentation on how to use logging, i. e. names of the loggers to use?

Comment (sils1297) [inline coalib/coala_main.py]: the you should add a constant for the filenme in Constants and use the calib root constant for getting the filename

Comment (sils1297) [inline coalib/coala_main.py]: as for the docs: not needed if its not for the user

File: coalib/output/printers/LogPrinter.py (hunk 13 above)
Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: How to know if the output is piped somewhere?
IMO, when you're writing something, you know whether it will be piped or no, so you can use what will be the most appropriate for you.

Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: I don't know but it's not something you should decide when getting a logger but the logger should decide it on its own

File: coalib/bears/Bear.py (hunk 9 above)
Comment (Uran198) [inline coalib/bears/Bear.py]: The usecases that I encountered worked with classes only, so I think it was the original purpose. I think you cannot have two methods with the same name, one as a class method and the other as regular one.

Comment (sils1297) [inline coalib/bears/Bear.py]: if the JSONEncoder has an object you can simply make a normal method and be fine. Or we could patch the JSONEncoder to do the getattr on its own.

File: coalib/output/printers/LogPrinter.py (hunk 13 above)
Comment (Uran198) [inline coalib/output/printers/LogPrinter.py]: The logger has much less context. Also, as it is right now, it always writes to sys.stdout, so it cannot get pipped. I'll configure list log printer, and it will not be used with colours, and can be used if you want something to be piped.

One idea here, is to configure loggers to be called as `getLogger(__name__)`. Then configure the respective "module level" loggers, this way it can decide which kind of logger needs to be used based on the module it is in.

File: coalib/bears/Bear.py (hunk 9 above)
Comment (Uran198) [inline coalib/bears/Bear.py]: Basically as it is right now, it is working the way as before, if there was an instance of the object passed to JSONEncoder, we would get value from `get_public_member` and we will not filter that out. Basically what you're saying is to make json a regular method, or to modify JSONEncoder. This I think is more a separate issue. Here I only correct the current behaviour, and you want to introduce a new one.

Comment (arafsheikh) [inline coalib/bears/Bear.py]: @Uran198 If `_dict` doesn't contain any property object then [this](https://github.com/coala-analyzer/coala/pull/2409/files#diff-7b888ece875aa379bf8cfcebcd7cdef1R282) should be removed as well. As it was there only to filter out un-serializable properties.

Comment (Uran198) [inline coalib/bears/Bear.py]: Nice catch @arafsheikh. I did not notice that it was hard coded before.

File: coalib/coala.py [hunk 15]
```diff
@@ -27,6 +30,10 @@
 
 
 def main():
+    # Configure logging
```
Comment (sils1297) [inline coalib/coala.py]: how about `# Loads logging configuration from JSON`?

File: coalib/coala.py [hunk 16]
```diff
@@ -27,6 +30,10 @@
...
+    with open(log_settings_file) as fd:
```
Comment (sils1297) [inline coalib/coala.py]: fd isn't a good variable name

File: coalib/coala.py [hunk 17]
```diff
@@ -27,6 +30,10 @@
...
+        logging.config.dictConfig(json.load(fd))
```
Comment (sils1297) [inline coalib/coala.py]: I feel like this thing belongs into a function, it's going to be needed in other main functions as well right?

File: tests/output/ConsoleInteractionTest.py [hunk 18]
```diff
@@ -3,9 +3,9 @@
 from unittest.mock import patch
 from collections import OrderedDict
 from os.path import abspath, relpath
+import logging
```
Comment (sils1297) [inline tests/output/ConsoleInteractionTest.py]: typo in commit message

File: tests/output/ConsoleInteractionTest.py [hunk 19]
```diff
@@ -758,8 +758,7 @@ def test_show_bears_capabilities(self):
 class PrintFormattedResultsTest(unittest.TestCase):
 
     def setUp(self):
-        self.printer = StringPrinter()
-        self.logger = LogPrinter(self.printer)
+        self.logger = LogPrinter()
```
Comment (sils1297) [inline tests/output/ConsoleInteractionTest.py]: is th logger only used below? Then it doesn't need to appear in setup.

File: coalib/coala_delete_orig.py [hunk 20]
```diff
@@ -7,9 +9,14 @@
 from coalib.settings.ConfigurationGathering import get_config_directory
 from coalib.settings.Section import Section
 from coalib.parsing.Globbing import glob_escape
+from coalib.misc.Constants import log_settings_file
 
 
 def main(log_printer=None, section: Section=None):
+    # Configure logging
+    with open(log_settings_file) as fd:
+        logging.config.dictConfig(json.load(fd))
```
Comment (sils1297) [inline coalib/coala_delete_orig.py]: yeah make it a function :) or a decorator? like `@configure_logging(config_file_path)`?

File: coalib/logging_config.json [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        "colored": {
+            "class": "logging.StreamHandler",
+            "formatter": "raw",
+            "stream": "ext://sys.stdout"
+        },
+        "raw": {
+            "class": "logging.StreamHandler",
+            "formatter": "raw",
+            "stream": "ext://sys.stdout"
+        }
```
Comment (sils1297) [inline coalib/logging_config.json]: hm those two handlers look pretty similar to each other... the same. How does the colored one get the color?

File: coalib/logging_config.json [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+        },
+        "coala.colored": {
+            "handlers": ["colored"]
+        }
+    },
+    "formatters": {
+        "raw": {
+            "format": "[%(levelname)s][%(asctime)s] %(message)s",
+            "datefmt": "%X"
+        }
```
Comment (sils1297) [inline coalib/logging_config.json]: no color formatter?

File: coalib/output/printers/LogPrinter.py [hunk 23]
```diff
@@ -29,26 +34,38 @@ def __init__(self,
...
+        self.logger = logging.getLogger(self.logger_name)
+
         self._printer = printer
         self.log_level = log_level
         self.timestamp_format = timestamp_format
 
     @property
+    def log_level(self):
+        """
+        Returns current log_level used is logger.
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: grammar

File: coalib/output/printers/LogPrinter.py [hunk 24]
```diff
@@ -29,26 +34,38 @@ def __init__(self,
...
+        """
+        return self._log_level
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: can't you get the log level from the logger directly?

File: tests/collecting/CollectorsTest.py [hunk 25]
```diff
@@ -202,20 +200,18 @@ def setUp(self):
...
-                                           self.log_printer), ([],))
-            self.assertRegex(sio.getvalue(),
-                             ".*\\[WARNING\\].*No bears were found matching "
-                             "'invalid_name'.\n")
+        self.assertEqual(collect_bears(["invalid_paths"],
+                                       ["invalid_name"],
+                                       ["invalid kind"],
+                                       self.log_printer), ([],))
+        self.assertEqual(''.join(log.message for log in self.log_printer.logs),
+                         "No bears were found matching 'invalid_name'.")
```
Comment (sils1297) [inline tests/collecting/CollectorsTest.py]: how about `self.assertEqual([log.message for log in self.log_printer], ['No ...'])`?

File: tests/processes/LogPrinterThreadTest.py [hunk 26]
```diff
@@ -1,12 +1,13 @@
...
 from coalib.misc.ContextManagers import retrieve_stdout
 from coalib.output.printers.LogPrinter import LogPrinter
 from coalib.processes.LogPrinterThread import LogPrinterThread
 
 
-class TestPrinter(LogPrinter):
+class TestPrinter(Printer, LogPrinter):
 
     def __init__(self):
         LogPrinter.__init__(self, self)
```
Comment (sils1297) [inline tests/processes/LogPrinterThreadTest.py]: you should call the init methods of all superclasses!

File: tests/processes/ProcessingTest.py [hunk 27]
```diff
@@ -51,7 +52,7 @@ def is_alive(self):
         return not self.control_queue.empty() and not self.starts_dead
 
 
-class ProcessingTestLogPrinter(LogPrinter):
+class ProcessingTestLogPrinter(Printer, LogPrinter):
 
     def __init__(self, log_queue):
         LogPrinter.__init__(self, self)
```
Comment (sils1297) [inline tests/processes/ProcessingTest.py]: same

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/f507e6cd3e9a126c9d4c386f61b94373e7aa5649 https://github.com/coala-analyzer/coala/pull/2409/commits/26aa4a5f624d7da2e749de94d9c564fa370d8c3e

File: coalib/logging_config.json [hunk 28]
```diff
@@ -3,7 +3,7 @@
     "handlers": {
         "colored": {
             "class": "logging.StreamHandler",
-            "formatter": "raw",
+            "formatter": "colored",
```
Comment (sils1297) [inline coalib/logging_config.json]: yeah the last commit makes more sense if this is squashed into i

File: tests/output/ConsoleInteractionTest.py (hunk 19 above)
Comment (Uran198) [inline tests/output/ConsoleInteractionTest.py]: It's used in almost every test case.

File: tests/processes/LogPrinterThreadTest.py (hunk 26 above)
Comment (Uran198) [inline tests/processes/LogPrinterThreadTest.py]: Actually, there is no need in these changes anymore. :)

**Comment by sils1297** [comment]
did you read my comment about squashing https://github.com/coala-analyzer/coala/pull/2409/commits/3e1708bd5bc6ebc13c3d8fb9dfa120494f15895e ?

**Comment by Uran198** [comment]
reack 377bf5b 2f90389

**Comment by Uran198** [comment]
@sils1297 Sorry, I've misinterpreted that one. Why do you want it squashed? Previous commit is already very big and this change makes sense to keep separate, though, I changed it a little, so right now it should make more sense. One possibility to make more logical can be to add `coala.colored` handler with colour formatter.

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/3454d2fc19aed8fe78b4f6c7adef4faa487a1374

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/30b9593f2a13c887e274a0687a502b5a871897ae

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/6820399c9c8bbbbc14134d6e00fe83a01a31d320

**Comment by sils1297** [comment]
@Uran198 can you make an asciinema ? How does the new logging look like? Also, is there a way to tell it to pause logging and resume it later, flushing messages that came in during that time?

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/6901813cec6a1239bf651d7d7a74165c87a68ea9

**Comment by sils1297** [comment]
maybe we want to wait with merging this until the release for stability?

**Comment by Uran198** [comment]
@sils1297 Basically, it looks exactly the same. Cause of that I only made asciinema to show that it works: https://asciinema.org/a/3fkds56dxj5fm6qpvrxi860rw . But it would be good if someone else would validate that it is working as well, as I may have forgotten to push some dependencies or something.

There is possibility not to flush messages right away, take a look at [MemoryHandler](https://docs.python.org/3/library/logging.handlers.html#memoryhandler), though it has limited capacity and flushes when it's full. But if needed there will be no problem in inheriting it and overriding default behavior to needed.

It's probably better to wait for a release before merging, IMO.

**Comment by Uran198** [comment]
reack 26ff23ecbf7e9b8af2b0c85f580cbab7527062ea 85500050a57ce1064453ad8dc12fa63593874370  ac25b1cc3eaddd955051963513f8c0ba77407888  5bd4e21b0d9e054d619cae3d85109194be44d6b5 ae69a9595efd1e974731139158f05d645407f933 49b8c5d2974d1417f63e3b5e8853b8531fde57ac

**Comment by sils1297** [comment]
@Uran198 I don't see how this would break our release, my opinion: let's just rebase and merge it now so we can profit from it.

**Comment by Uran198** [comment]
reack 410ace8 fb09f49 c90f25d 0ef2d65 9a86ed5

**Comment by Uran198** [comment]
reack cd2b8c4

File: tests/output/printers/LogPrinterTest.py [hunk 29]
```diff
@@ -66,12 +65,8 @@ def test_set_state(self):
         uut = LogPrinter()
         state = uut.__getstate__()
         uut.__setstate__(state)
-        self.assertEqual(uut.logger, logging.getLogger('coala.raw'))
-
-    def test_colored_printer(self):
-        uut = LogPrinter(printer=ColorPrinter())
-        self.assertIs(uut.logger, logging.getLogger('coala.colored'))
+        self.assertEqual(uut.logger, logging.getLogger('coalib'))
```
Comment (hypothesist) [inline tests/output/printers/LogPrinterTest.py]: Why would this be `self.assertEqual` but the other be `self.assertIs`? IIUC both should be `self.assertIs`?

**Comment by hypothesist** [comment]
So https://github.com/coala-analyzer/coala/pull/2409/commits/f206732a312d99c0038ed40f7b613407e91d60e6 looks alright to me, except for that one question

**Comment by Uran198** [comment]
reack 40dedbe716dd3ea82ad1c2784c0039436b789589

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/d84184b223369c4d45c0fe96eaf89ec50b4cf796

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/4b223ae2471a675ece71690acf49ab5721e6a2a3

**Comment by sils1297** [comment]
hm @Uran198 when I run coala with your PR it doesn't show debug log messages of bears it seems

**Comment by Uran198** [comment]
@sils1297 This will require a lot of testing and may break a lot of things, cause LogPrinter is created in a lot of places, and log_level can be changed in any one of them. I'm thinking of switching already to the `logging.getLogget(__name__)` style, where this happens. This probably should go into d84184b commit, which will bloat it even more. Another alternative, is to go to the compromise and have intermediate version where it's working, but buggy.

**Comment by Uran198** [comment]
reack 0f9c8c0 e68cefa 7ce0fcb b0972e1 bbf7281 4e3bd71 417c40e

**Comment by Uran198** [comment]
@sils1297 Could you verify that it still doesn't show debug messages? Make sure that the logging level is set up correctly, as by default it is info, so it may have been correct behavior when it did not show debug messages.

**Comment by sils1297** [comment]
now it always shows debug messages

**Comment by sils1297** [comment]
even if I run coala -L WARNING

**Comment by Uran198** [comment]
@sils1297 This is really strange cause I did not change the original code, only rebased it.

**Comment by sils1297** [comment]
Unable to reproduce my issues in a clean venv

**Comment by Uran198** [comment]
reack 6e962b321f7e6e4fb7ffc8e7ceb6814ece0c9fd9 048573274dceb873917919af48c1b322f78ada06 f6367751ed286e4cff891cabedbdb1a9f37c31e8 189e6c7027184b229af1a842a6a5ada989d631c7 7dd39e1e5b3a9c34def41b941e602fc0b04bba8e 718b7289788d69bb66b4a45c38e6b6d1601b086b 7a3532b796aa3fbd49e3723a16929ba20afee35a

**Comment by Uran198** [comment]
reack 7fd00141c98e7d003ec524538c6fd25e2c5aa626 c18996e57fc1518fae061b4bf278fe8e97f089b0 8c261cef9f1a652371a2c20d4a3d21ea30c165e4 7fb344c19abcc42926feeb3e74838f2f44e4cb8e 08d0d76f61b5894ab02d3fe0442b6fb67320e1d1 6aba75fada2a0a0f257fbcc526cd422e64523d2b 96972f2249b1b8b47060256f398489bb54c230c9

**Comment by Uran198** [comment]
There were some changes in 8c261ce, but in got reacked anyway. So this one also need reviewing.

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/commit/8c261cef9f1a652371a2c20d4a3d21ea30c165e4 then as it needs review

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2409/commits/96972f2249b1b8b47060256f398489bb54c230c9

File: coalib/settings/ConfigurationGathering.py [hunk 30]
```diff
@@ -122,6 +124,16 @@ def warn_config_absent(sections, argument, log_printer):
...
 
 
+def configure_logging_from_json(filename):
+    """
+    Configures logging module based on parameters from JSON file.
+
+    :param filename:    Name of a file with logging configuration.
+    """
+    with open(filename) as cnf_file:
+        logging.config.dictConfig(json.load(cnf_file))
```
Comment (underyx) [inline coalib/settings/ConfigurationGathering.py]: What's the benefit of having a separate JSON file instead of just keeping the config somewhere as a Python dictionary?

Comment (Uran198) [inline coalib/settings/ConfigurationGathering.py]: Separation of concerns. When settings are separated, it's easier to find
and edit them. In particular, if you will want to add, say, the ability for
users to have their own logging settings, it will much easier to change
from configuration in the file. What are your thoughts? Do you think it'll
be better to have it somewhere in code?

Comment (sils1297) [inline coalib/settings/ConfigurationGathering.py]: I think putting it in constants is simpler and also better for performance.
It looks same...

Comment (underyx) [inline coalib/settings/ConfigurationGathering.py]: The separation of concerns requirement can be satisfied by just moving this to its own module. That would still save us the trouble of
- Managing files in the code manually
- Having to bring in JSON serialization
- Having to write JS code

But anyway, I don't think 'separation of concerns' applies at the level. Loading configuration (`ConfigurationGathering.py`) and logging configuration (`logging_config.json`) are the same concern, so it might be okay. The issue would be if the config gatherer suddenly started installing bears that are mentioned in the config file or something :)

**Comment by sils1297** [comment]
oh no conflicts :/

**Comment by sils1297** [comment]
can you rebase? I think it would be good to move the config into code as well.

**Comment by Uran198** [comment]
reack a218bce 7ccd6b2 eb7859e 80dca05

**Comment by sils1297** [comment]
ack https://github.com/coala/coala/pull/2409/commits/578fe472bb3acc50e6ea98bd68d30945ce18fc03#diff-f9283b08ae99b5a2f6d1781935a9b8d5L10

File: coalib/misc/Constants.py [hunk 31]
```diff
@@ -90,3 +90,31 @@
...
+                'handlers': ['raw']
+            }
+        },
+        'formatters': {
+            'raw': {
+                'format': '[%(levelname)s][%(asctime)s] %(message)s',
+                'datefmt': '%X'
+            }
+        }
+    }
```
Comment (sils1297) [inline coalib/misc/Constants.py]: why is this a function and not a constant?

File: coalib/output/printers/LogPrinter.py [hunk 32]
```diff
@@ -1,6 +1,8 @@
 import traceback
 import logging
 
+from pyprint.ColorPrinter import ColorPrinter
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: isn't that a recursive import then?

**Comment by sils1297** [comment]
ack https://github.com/coala/coala/pull/2409/commits/17155e3985a2ed8a2c8c115469cb467d9f56250b - beautiful solution!

File: coalib/output/printers/LogPrinter.py [hunk 33]
```diff
@@ -117,7 +120,10 @@ def __init__(self,
         :param timestamp_format: The format string for the
                                  datetime.today().strftime(format) method.
         """
-        self.logger_name = 'coala.raw'
+        if isinstance(printer, ColorPrinter):
+            self.logger_name = 'coala.colored'
+        else:
+            self.logger_name = 'coala.raw'
```
Comment (sils1297) [inline coalib/output/printers/LogPrinter.py]: this is a bad design pattern, ideally the ColorPrinter does what it needs on its own and this LogPrinter shouldn't have anything to do with it

File: coalib/misc/Constants.py (hunk 31 above)
Comment (sils1297) [inline coalib/misc/Constants.py]: Sys is usually used anyway, problems are imports from non builtins

**Comment by sils1297** [comment]
@Uran198 you left my comments unanswered

**Comment by sils1297** [comment]
besides those it looks good

**Comment by Uran198** [comment]
@sils1297 sorry for not responding, I agree with your comments but had no time to fix it. And I just forgot to label this WIP.

File: coalib/misc/Constants.py (hunk 31 above)
Comment (Uran198) [inline coalib/misc/Constants.py]: I'll add the comment for this.

**Comment by Uran198** [comment]
reack c48944e 98fb89d 149e322 6f27f70

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2409/commits/6f27f70737f3f50056f5f9c0c86ccdf03cf7f93f

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2409/commits/6f27f70737f3f50056f5f9c0c86ccdf03cf7f93f

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2409/commits/f41c4179fbbd2fd14ad285963e21289dde2a0571

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2409/commits/bc515e990712456b1f7a17246e7a6a451ed1f0e5

File: coalib/coala.py [hunk 34]
```diff
@@ -27,6 +29,8 @@
 
 
 def main():
+    logging.config.dictConfig(get_log_config_dict())
```
Comment (sils) [inline coalib/coala.py]: maybe that function should actually do the dictConfig invocation as well - less duplication cross all those files

File: tests/coalaCITest.py [hunk 35]
```diff
@@ -60,6 +60,7 @@ def test_fix_patchable_issues(self):
                 "-b", "SpaceConsistencyTestBear",
                 "--settings", "autoapply=true", "use_spaces=True",
                 "default_actions=SpaceConsistencyTestBear:ApplyPatchAction")
+            print(output)
```
Comment (sils) [inline tests/coalaCITest.py]: debug print?

**Comment by sils** [comment]
can't find anything but those

**Comment by Uran198** [comment]
reack 0312153809527d29358304d7cba7091c6005d381 2646d7e658dfc421f95cc0f9cae347306c5de952 8442f0375964c2c5b97c3f66757ded5a3557aac7 b4c4e04eee149f3f392e7ee9c7c00cfcb4e3a62b b9b8a9d25ccadaea6d55da976c305ce19a357a4c d5c0c3912282bf26cfa5e24dd9b17e9355c791ef

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2409/commits/d5c0c3912282bf26cfa5e24dd9b17e9355c791ef

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/2409/commits/c657d44b62e8fac62c2504c96f6848f302a11c35

**Comment by sils** [comment]
@rultor merge

**Comment by sils** [comment]
@Uran198 can you file issues about what needs to be done for the full switch and maybe close the original one or make it a master issue linking to the others?

**Comment by sils** [comment]
Also this is great work :)
